### PR TITLE
Dagger category theory

### PR DESCRIPTION
**Pull request by Kfwullaert**
This PR is the start of the formalization of theory of dagger categories.
The following material is currently provided:
- Definitions of dagger categories, dagger functors, dagger adjoint, dagger isos and univalence
- Definitions of weak and strong dagger equivalences between dagger categories (no results, only definitions)
- A proof that any fully faithful functor into, a dagger category, 'creates' a unique dagger structure on the domain such that the functor is a dagger functor
- A construction of the functor-dagger category and a proof that it is dagger-univalent if the codomain is dagger-univalent.

Furthermore, two examples of dagger categories together with a univalence result:
1. The category of relations: This category is introduced, it is shown that this is both univalent and dagger univalent.
2. Any groupoid becomes a dagger category and a proof that it is dagger univalent if and only if it is univalent.

### CODE REVIEW

**Review by m-lindgren — COMMENTED** [review]
Looks like good material. I'll look at it more closely tomorrow. 

Maybe the `Core` folder isn't necessary? The imports get very long currently. What do you think about having the files currently in `Core` directly in `DaggerCategories` instead, and renaming them `Categories.v`, `Functors.v` etc?

File: UniMath/CategoryTheory/DisplayedCats/Examples/Relations.v
Comment (Kfwullaert) [reply UniMath/CategoryTheory/DisplayedCats/Examples/Relations.v:211]: Thanks a lot for this, this will make the code a lot nicer.

File: UniMath/CategoryTheory/DisplayedCats/Examples/Relations.v
Comment (Kfwullaert) [reply UniMath/CategoryTheory/DisplayedCats/Examples/Relations.v:26]: I have missed that for some reason. If I have a proposal for a better name, I will post it here before making any changes.

File: UniMath/CategoryTheory/DisplayedCats/Examples/Relations.v
Comment (nmvdw) [inline UniMath/CategoryTheory/DisplayedCats/Examples/Relations.v:1]: Why is this file in this folder? It doesn't seem to be a displayed category, or am I missing something?

File: UniMath/CategoryTheory/DaggerCategories/Examples/groupoids.v
Comment (nmvdw) [inline UniMath/CategoryTheory/DaggerCategories/Examples/groupoids.v:1]: Can you call this file Groupoids.v (with a capital)? The same for `relations.v` in DaggerCategories.

**Review by nmvdw — COMMENTED** [review]
(Only some small comments on file names and locations)

**Comment by nmvdw** [comment]
> > Looks like good material. I'll look at it more closely tomorrow.
> > Maybe the `Core` folder isn't necessary? The imports get very long currently. What do you think about having the files currently in `Core` directly in `DaggerCategories` instead, and renaming them `Categories.v`, `Functors.v` etc?
> 
> This is now done.

You missed CONTENTS.md

File: UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v
Comment (m-lindgren) [inline UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v:58]: I'm surprised we don't have this in the library already, at least specific to composition of morphisms in a category. 

The general case would be `maponpaths_12` I think, and for composition it would be `maponpaths_12 compose`. In fact I suspect there are many instances of `etrans; apply maponpaths_12` that could/should be replaced with this.

But this seems like a useful function to have since it's the general case of `cancel_postcomposition` and such, and I would be strongly in favor of upstreaming this.

File: UniMath/CategoryTheory/DaggerCategories/Examples/Groupoids.v
Comment (m-lindgren) [inline UniMath/CategoryTheory/DaggerCategories/Examples/Groupoids.v:21]: I think this already exists. `inverse_unique_precat` seems to have the correct type.

**Review by m-lindgren — CHANGES_REQUESTED** [review]
Not quite done with the review yet, but only some minor comments so far.

File: UniMath/CategoryTheory/DaggerCategories/Examples/Groupoids.v
Comment (benediktahrens) [reply UniMath/CategoryTheory/DaggerCategories/Examples/Groupoids.v:49]: Yes, I strongly second this suggestion.

File: UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v
Comment (benediktahrens) [reply UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v:58]: Two or three years ago, @peterlefanulumsdaine replaced many uses of `cancel_postcomposition` by uses of `maponpaths_12`. I don’t have a strong opinion either way, but we should not spend effort switching between the general and the specialized lemma.

File: UniMath/CategoryTheory/DaggerCategories/Examples/Groupoids.v
Comment (Kfwullaert) [reply UniMath/CategoryTheory/DaggerCategories/Examples/Groupoids.v:49]: There is already a lemma available: groupoid_is_pregroupoid.
I should have looked more at the code in groupoids.v.

File: UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v
Comment (peterlefanulumsdaine) [reply UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v:58]: @m-lindgren : What was your reason for preferring this specific lemma over the general `mapopaths_12`?  I find the general lemmas almost always just as easy to apply (`maponpaths_12` for both arguments at once, or `maponpaths`/`maponpaths_2` for one argument at a time), so I don’t see the advantage of the specialised one.

Overall I have a mild preference for using the general lemma (I don’t object to other people use a specialised one, but I prefer to use the general myself unless there are compelling reasons otherwise) — but I have a very strong preference for keeping their names unified as some variant of `maponpaths`.  The point is that `maponpaths` and its specialisations are a pieces of completely general, mathematically near-trivial algebraic reasoning, that applies to all functions; so (a) that trivialness should be kept as clear as possible, (b) people shouldn’t have to learn a different set of lemma names for algebraic reasoning with different operations, and (c) non-unified names means people don’t discover it, so they re-invent it under different names.  When I replaced a lot of instances by `maponpaths_12` a few years ago, it was because there were about half a dozen different copies of this lemma under different names scattered through the library.  I don’t mind having one specialised version, but I’m strongly against having multiple ones.  So if we keep this, please could we rename it to `maponpaths_compose` or similar?

Comment (m-lindgren) [reply UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v:58]: @Kfwullaert Yes, that sounds good, and maybe write the proof using maponpaths_12 so that it's obvious that it is an intended specialization of the general lemma?

@benediktahrens Ah, interesting! In this case I think it fits in well with the current use of `cancel_postcomposition` etc and would be a good addition for new code entering CategoryTheory, but not something to spend time on changing old code.

Comment (m-lindgren) [reply UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v:58]: @peterlefanulumsdaine So for one, this lemma is opaque and `maponpaths_12` is not. Another reason is discoverability: This lemma is domain specific and expressed using the notation that is common when working with categories, and it's input is explicitly morphisms of a category, making it easy to find. I think of this as providing an API to working with/in categories, where the underlying structure and implementation is abstracted away and these helper functions provides a nice interface to use, which also provides somewhat more readable code.

Comment (m-lindgren) [reply UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v:58]: @peterlefanulumsdaine I have also seen duplications of the general lemma, before your cleanup, and I would of course strongly object to anyone adding more duplication like that. :-) But instances like this one is not duplication, it's specialization for the sake of readability, usability and discoverability, IMO.

File: UniMath/CategoryTheory/DaggerCategories/Univalence.v
Comment (m-lindgren) [inline UniMath/CategoryTheory/DaggerCategories/Univalence.v:66]: Maybe have a coercion so these (and similar) `pr1` can be removed? Or at least have an accessor function with a descriptive name.

I think in particular in type declarations one should avoid using `prX`, unless absolutely necessary.

File: UniMath/CategoryTheory/DaggerCategories/Univalence.v
Comment (m-lindgren) [inline UniMath/CategoryTheory/DaggerCategories/Univalence.v:82]: Same as above, but in this case I think you can just turn the identity around so you have `dagC x y (Univalence.idtoiso p)) = ...`, and the coercions should trigger.

Also, maybe import Univalence?

File: UniMath/CategoryTheory/categories/Relations.v
Comment (m-lindgren) [inline UniMath/CategoryTheory/categories/Relations.v:34]: Just reminding my self that `hrel` needs another name, and just noticed that `eq_set` here is a duplicate of `eqset` defined in Foundations.Sets.

File: UniMath/CategoryTheory/Core/Categories.v
Comment (m-lindgren) [inline UniMath/CategoryTheory/Core/Categories.v:284]: Actually what I liked about this function the most was the name you gave it before, it was very category theory-specific, which I think will promote readable code.

File: UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v
Comment (m-lindgren) [inline UniMath/CategoryTheory/DaggerCategories/FunctorCategory.v:351]: Maybe replace pr1 with accessors or suitable coercion?

**Review by m-lindgren — CHANGES_REQUESTED** [review]
Overall nice material, I learned things i didn't know about dagger categories reviewing this. I'm mostly concerned about finding a better name for `hrel`, the rest is style comments.

File: UniMath/CategoryTheory/DaggerCategories/Univalence.v
Comment (Kfwullaert) [reply UniMath/CategoryTheory/DaggerCategories/Univalence.v:66]: In this case, you mean something like idtomor (in the non-dagger/just categories case)?

File: UniMath/CategoryTheory/DaggerCategories/Univalence.v
Comment (Kfwullaert) [reply UniMath/CategoryTheory/DaggerCategories/Univalence.v:82]: That would indeed be nicer, not sure why I chose this form.
Regarding the import of Univalence, I should have mentioned this as a comment in this PR (and even in the file),
this time it was not that I forgot to include it. I choose to not include it because you are not interested in univalence, but in dagger univalence and therefore, I think it is better practice to not include it. I'm aware this not standard, so I can change this.
The (only) reason I have used Univalence.idtoiso is because this form allows me to use lemmas of about idtoiso that have are already been formalized. I could avoid this and redefine it for idtodaggeriso. However, a lot things had to defined yet again, very analogous to the non-dagger case. So I used it more as a 'trick', but strictly speaking, it is not necessary.

File: UniMath/CategoryTheory/DaggerCategories/Unitary.v
Comment (Kfwullaert) [reply UniMath/CategoryTheory/DaggerCategories/Unitary.v:82]: I could replace them by {} (instead of 1: ),
e.g. etrans. { apply  assoc. }

**Comment by Kfwullaert** [comment]
@m-lindgren Regarding the name "hrel". I have no "good" idea, but I can only think of hrel_2. The 2 denotes that it is a relation between 2 sets. So, hrel, would be considered to be hrel_1 as it is a relation between 1 set.

**Comment by benediktahrens** [comment]
> @m-lindgren Regarding the name "hrel". I have no "good" idea, but I can only think of hrel_2. The 2 denotes that it is a relation between 2 sets. So, hrel, would be considered to be hrel_1 as it is a relation between 1 set.

I am not seeing the context of this discussion, but a relation between two things is sometimes called a "binary relation". Maybe that could be used to find a suitable name?

**Comment by Kfwullaert** [comment]
> > @m-lindgren Regarding the name "hrel". I have no "good" idea, but I can only think of hrel_2. The 2 denotes that it is a relation between 2 sets. So, hrel, would be considered to be hrel_1 as it is a relation between 1 set.
> 
> I am not seeing the context of this discussion, but a relation between two things is sometimes called a "binary relation". Maybe that could be used to find a suitable name?

@benediktahrens The context of this discussion is as follows:
Currently, there is already another definition of hrel. But this is given as the type of relations on X, i.e. hrel X := X -> X -> hProp.
But the definition of hrel that I've now provided, is hrel X Y := X -> Y -> hProp.
Thus, the problem currently is that I've introduced a type whose name already exists. Therefore, I have to change the name of the latter.

**Comment by benediktahrens** [comment]
Mmh, possibly the type of the original `hrel` should be generalized - I don’t see the benefit of restricting to the special case of the two types being the same. We can discuss this at the UniMath User Meeting.

For the purpose of this PR, I suggest you use `bin_hrel` or similar.

**Comment by Kfwullaert** [comment]
I think I have gone through every comment.
What I think currently remains (or at least of those parts that already have been reviewed), is the decision of the name for the above mentioned lemma (f1=f2 -> g1=g2 -> f1g1 =f2g2). Currently it is maponpaths_compose (as proposed by @peterlefanulumsdaine), but @m-lindgren prefers equality_of_composition.

**Comment by rmatthes** [comment]
I find both names well justified. Not helping with the decision...

**Comment by m-lindgren** [comment]
I'm fine with merging this as is, having dagger categories in the library will be nice. By the way, @Kfwullaert any chance we'll see monoidal  dagger categories as well?

**Comment by Kfwullaert** [comment]
> I'm fine with merging this as is, having dagger categories in the library will be nice. By the way, @Kfwullaert any chance we'll see monoidal dagger categories as well?

Yes, but before starting, I wait for PR #1644 to be merged.

**Pull request merged**
